src/security/noise/crypto/noise_dh.cpp

```diff
@@ -0,0 +1,37 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <libp2p/crypto/error.hpp>
+#include <libp2p/security/noise/crypto/noise_dh.hpp>
+
+namespace libp2p::security::noise {
+
+  outcome::result<DHKey> DiffieHellmanImpl::generate() {
+    OUTCOME_TRY(keypair, x25519.generate());
+    ByteArray priv{keypair.private_key.begin(), keypair.private_key.end()};
+    ByteArray pub{keypair.public_key.begin(), keypair.public_key.end()};
+    return DHKey{.priv = std::move(priv), .pub = std::move(pub)};
+  }
+
+  outcome::result<ByteArray> DiffieHellmanImpl::dh(
+      const ByteArray &private_key, const ByteArray &public_key) {
+    crypto::x25519::PrivateKey priv;
+    crypto::x25519::PublicKey pub;
+    if (private_key.size() != priv.size() or public_key.size() != pub.size()) {
+      return crypto::OpenSslError::WRONG_KEY_SIZE;
+    }
+    std::copy_n(private_key.begin(), priv.size(), priv.begin());
+    std::copy_n(public_key.begin(), pub.size(), pub.begin());
+    return x25519.dh(priv, pub);
+  }
+
+  int DiffieHellmanImpl::dhSize() const {
+    return 32;
+  }
+
+  std::string DiffieHellmanImpl::dhName() const {
+    return "25519";
```

Comment:
Why not x25519, is this from noise spec?

Reply:
I'm just concerned that there's a plenty of things called *25519.

Reply:
The constant is prescribed by noise spec. https://github.com/libp2p/specs/tree/master/noise#noise-protocol-name
<img width="1044" alt="image" src="https://user-images.githubusercontent.com/36847043/96260920-5cd0dd80-0fc8-11eb-940c-0aae2f498966.png">

Note that XX here stands for handshake pattern name